### PR TITLE
 fixing issue #116  cleanup fail in case of exception in nbadapter.py:429

### PR DESCRIPTION
**Pull request by okolo**
Below is an example of runtime error when running test_CRbeam.ipynb from renkulab:
https://renkulab.io/projects/astronomy/mmoda/crbeam

```python
File /opt/conda/lib/python3.10/site-packages/nb2workflow/nbadapter.py:403, in NotebookAdapter.execute(self, parameters, progress_bar, log_output, inplace)
400 logstasher.log(dict(origin="nb2workflow.execute", event="starting", parameters=parameters, workflow_name=notebook_short_name(self.notebook_fn), health=current_health()))
402 logger.info("starting job")
--> 403 exceptions = self._execute(parameters, progress_bar, log_output, inplace)
405 tspent = time.time() - t0
406 logstasher.log(dict(origin="nb2workflow.execute",
407 event="done",
408 parameters=parameters,
(...)
411 health=current_health(),
412 time_spent=tspent))

File /opt/conda/lib/python3.10/site-packages/nb2workflow/nbadapter.py:429, in NotebookAdapter._execute(self, parameters, progress_bar, log_output, inplace)
426 except Exception as e:
427 logger.warning("git clone failed: %s, will attempt copytree", e)
--> 429 os.rmdir(tmpdir)
431 shutil.copytree(os.path.dirname(os.path.realpath(self.notebook_fn)), tmpdir)
432 else:

OSError: [Errno 39] Directory not empty: '/tmp/nb2w-bn8qff1q'
```

### CODE REVIEW

**Comment by volodymyrss** [comment]
Now I remember what is this about, `rmdir` was sufficient since the failure cloning should not have created any files. The tmp directly was just created few lines ago, so it should be empty. `rmtree` is kind of scary so it's best avoided if not needed.

It seems like in this case git LFS has an issue with checkout but not clone which is why the directory is not empty.

```python
 base ▶ tmp ▶ $ ▶ git clone $HOME/work/crbeam
Cloning into 'crbeam'...
done.
Downloading data/0 (117 KB)
Error downloading object: data/0 (2cce9ed): Smudge error: Error downloading data/0 (2cce9ed48091a7cbfe090dbe03384a3505d0617a2f574f9890df22ff544ace5f): error transferring "2cce9ed48091a7cbfe090dbe03384a3505d0617a2f574f9890df22ff544ace5f": [0] remote missing object 2cce9ed48091a7cbfe090dbe03384a3505d0617a2f574f9890df22ff544ace5f

Errors logged to '/tmp/crbeam/.git/lfs/logs/20231016T113844.375144801.log'.
Use `git lfs logs last` to view the log.
error: external filter 'git-lfs filter-process' failed
fatal: data/0: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'
```

It might have something to do with some renku-specific behavior. We probably want to make a question to [renku.discourse](https://renku.discourse.group/)
I find some issues about this, e.g. https://github.com/git-lfs/git-lfs/issues/4095

It needs some investigation before we conclude that `rmtree` is the needed fallback.
Meanwhile, maybe we can make `rmtree` conditional on some `nbrun` argument?

**Comment by dsavchenko** [comment]
@okolo I just tried to start a clean session in renku with the latest master and there is no such error

**Comment by volodymyrss** [comment]
> @okolo I just tried to start a clean session in renku with the latest master and there is no such error

Maybe renku got upgraded or something? Could you try with a new session @okolo ?

**Comment by volodymyrss** [comment]
> > @okolo I just tried to start a clean session in renku with the latest master and there is no such error
> 
> Maybe renku got upgraded or something? Could you try with a new session @okolo ?

Could you confirm if this is still happening, @okolo ? Then, if you have indeed a problem cloning git repo inside renku session, we can report it to renku team.

**Review by volodymyrss — COMMENTED** [review]
need to see if this is renku issue, and check if it happens now

**Comment by okolo** [comment]
I have just checked that the problem can be reproduced on renku. Below are steps to reproduce:
- start new session of crbeam project https://renkulab.io/projects/astronomy/mmoda/crbeam/sessions
- run noteboook test_CRbeam.ipynb

**Comment by dsavchenko** [comment]
Now I understand.
In the session with "Automatically fetch LFS" everything works fine. Otherwise, this error occurs. Manually running `git lfs pull` before doing *nbrun* fixes it. 

As far as the file located in lfs is needed anyway, subsequent copytree won't help to properly execute the nb anyway. 
So this is not a renku-specific issue, it's just an issue which occurs in specific conditions when lfs files haven't been pulled. The exception should be raised anyway in these conditions, the only improvement could be to raise a more appropriate exception with error message, proposing to try `git lfs pull`. 

Or we can pull lfs in nbadapter before cloning the repo, but we need to think a bit, if this won't cause any side effects.

**Comment by volodymyrss** [comment]
> Now I understand. In the session with "Automatically fetch LFS" everything works fine. Otherwise, this error occurs. Manually running `git lfs pull` before doing _nbrun_ fixes it.
> 
> As far as the file located in lfs is needed anyway, subsequent copytree won't help to properly execute the nb anyway. So this is not a renku-specific issue, it's just an issue which occurs in specific conditions when lfs files haven't been pulled. The exception should be raised anyway in these conditions, the only improvement could be to raise a more appropriate exception with error message, proposing to try `git lfs pull`.
> 
> Or we can pull lfs in nbadapter before cloning the repo, but we need to think a bit, if this won't cause any side effects.

@okolo could you make the necessary implementation if this makes sense?

**Comment by volodymyrss** [comment]
If parent repository had no LFS, and git clone failed because of this, we should report the failure to the user with a suggestion to run git lfs ... command.

To detect this condition we can inspect captured Excepttion in more detail

**Comment by okolo** [comment]
> If parent repository had no LFS, and git clone failed because of this, we should report the failure to the user with a suggestion to run git lfs ... command.
> 
> To detect this condition we can inspect captured Exception in more detail. Check if called process error is useful, or parse command output and [find string](https://github.com/volodymyrss/pilton/blob/e1ac14d6a2dce7ec96f7ca72fc80d376941eacba/pilton.py#L376).

I have checked that git-lfs error may occur only if original repo was cloned with different version of git, e.g. if we run docker and mount the repo dir using -v option. Anyway the corresponding checks were added to the code.

File: nb2workflow/nbadapter.py [hunk 1]
```diff
@@ -431,21 +431,32 @@ def execute(self, parameters, progress_bar = True, log_output = True, inplace=Fa
...
+                    os.rmdir(tmpdir)
+                except OSError: # directory is not empty
+                    # this error may occure if the repo was originally cloned by the different version of git utility
+                    # e.g. when repo is mounted with docker run -v
+                    if 'git-lfs' in git_output:
+                        # the known solution is just to initialize git-lfs in advance
+                        message = "git-lfs is not initialized"
+                    else:
+                        message = "unknown git clone error, see logs for details"
+                    raise Exception(message)
```
Comment (burnout87) [inline nb2workflow/nbadapter.py:454]: Perhaps using the `GitPython` library you can better inspect the error returned from the clone command.

I used within the dispatcher, here a clone example:

https://github.com/oda-hub/dispatcher-app/blob/f8098edb7c19931fe41e4dc812814e27d54cfedd/cdci_data_analysis/analysis/renku_helper.py#L161C1-L175

**Review by burnout87 — COMMENTED** [review]
a suggestion about the git command

**Comment by okolo** [comment]
> a suggestion about the git command
I have implemented cloning using GitPython

File: tests/test_nbadapter.py [hunk 2]
```diff
@@ -123,6 +123,29 @@ def test_nbadapter_repo(test_notebook_repo):
...
+            os.remove(nba.output_notebook_fn)
+
+        if os.path.exists(nba.preproc_notebook_fn):
+            os.remove(nba.preproc_notebook_fn)
+
+        try:
+            nba.execute(dict())
+            assert False, 'nba.execute is expected to fail'
+        except Exception as ex:
+            assert ex.message == "git-lfs is not initialized"
```
Comment (volodymyrss) [inline tests/test_nbadapter.py:147]: wait, so is the test working, i.e. the failure successfully reproduced?

File: nb2workflow/nbadapter.py [hunk 3]
```diff
@@ -431,21 +431,25 @@ def execute(self, parameters, progress_bar = True, log_output = True, inplace=Fa
...
             try:
-                output = subprocess.check_output(["git","clone", "--recurse-submodules", os.path.dirname(os.path.realpath(self.notebook_fn)), tmpdir])
-                # output = subprocess.check_output(["git","clone", "--depth", "1", "file://" + os.path.dirname(os.path.realpath(self.notebook_fn)), tmpdir])
-                logger.info("git clone output: %s", output)
-            except Exception as e:
-                logger.warning("git clone failed: %s, will attempt copytree", e)
-
+                repo = Repo(repo_dir)
+                repo.clone(tmpdir, multi_options=["--recurse-submodules"])
+            except InvalidGitRepositoryError:
```
Comment (volodymyrss) [inline nb2workflow/nbadapter.py:441]: I am a little bit concerned about using an additional dependency (which is also just a wrapped around binary, and is not even actively maintained). So we avoided it in favor of direct git binary manipulation.
But if this exception handling is hard to implement right here, maybe it's justified.

File: tests/test_nbadapter.py (hunk 2 above)
Comment (okolo) [reply tests/test_nbadapter.py:147]: Unfortunately no. As you remember when we tried to reproduce the error we did not succeed unless we mount the existing git repository via -v option in docker provided that git versions inside docker and host machine are different. Even this effect depends on the particular version pairs. For this reason we decided to mark the test code as disabled.

File: nb2workflow/nbadapter.py (hunk 3 above)
Comment (okolo) [reply nb2workflow/nbadapter.py:441]: I partially agree. The only income which we get from the GitPython library usage is that it automatically recognizes the folders which do not contain repository and so we handle such folders in a more transparent way.

File: nb2workflow/nbadapter.py [hunk 4]
```diff
@@ -431,21 +431,25 @@ def execute(self, parameters, progress_bar = True, log_output = True, inplace=Fa
...
+                logger.warning(f"repository {repo_dir} is invalid, will attempt copytree")
                 os.rmdir(tmpdir)
-
                 shutil.copytree(os.path.dirname(os.path.realpath(self.notebook_fn)), tmpdir)
+            except GitCommandError as e:
+                logger.warning(f"git command error: {e}")
+                if 'git-lfs' in str(e):
```
Comment (burnout87) [inline nb2workflow/nbadapter.py:447]: just for my understanding, this is detect the missing of git-lfs right? perhaps we can use a more specific error message that is returned by git  ?

File: setup.py [hunk 5]
```diff
@@ -85,7 +85,8 @@
         'scrapbook', 
         'werkzeug==2.0.3',
         'sentry_sdk',
-        'rdflib'
+        'rdflib',
```
Comment (burnout87) [inline setup.py:88]: might be better to restring `rdflib` version following https://github.com/oda-hub/dispatcher-app/pull/646

File: nb2workflow/nbadapter.py (hunk 4 above)
Comment (volodymyrss) [reply nb2workflow/nbadapter.py:447]: Presumably this should refer to any git-lfs-related issue. Maybe the error message just below can be adjusted slightly.

File: nb2workflow/nbadapter.py [hunk 6]
```diff
@@ -431,21 +431,25 @@ def execute(self, parameters, progress_bar = True, log_output = True, inplace=Fa
...
+                    # this error may occure if the repo was originally cloned by the different version of git utility
+                    # e.g. when repo is mounted with docker run -v
+                    raise Exception("git-lfs is not initialized")
```
Comment (volodymyrss) [inline nb2workflow/nbadapter.py:450]:
```suggestion
                    raise Exception("We got some problem cloning the repository, the problem seems to be related to git-lfs. You might want to try reinitializing git-lfs with <.....>")
```

File: setup.py (hunk 5 above)
Comment (okolo) [reply setup.py:88]: This was done on master. I've merged the changes to this branch.

File: nb2workflow/nbadapter.py (hunk 6 above)
Comment (okolo) [reply nb2workflow/nbadapter.py]: Done

File: tests/test_nbadapter.py [hunk 7]
```diff
@@ -123,6 +123,29 @@ def test_nbadapter_repo(test_notebook_repo):
 
         validate_oda_dispatcher(nba)
 
+@pytest.mark.skip(reason="Not implemented")
```
Comment (dsavchenko) [inline tests/test_nbadapter.py:126]: Why is it skipped?

**Review by volodymyrss — APPROVED** [review]

File: tests/test_nbadapter.py (hunk 7 above)
Comment (dsavchenko) [reply tests/test_nbadapter.py:126]: Ok, let's keep it skipped, with the proposed reason message

**Review by dsavchenko — APPROVED** [review]

**Review by burnout87 — APPROVED** [review]

**Comment by dsavchenko** [comment]
So we all approved. Should we finally merge?

**Pull request merged**